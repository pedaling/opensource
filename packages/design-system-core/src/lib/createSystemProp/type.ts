import type { Theme } from '../types';

export type InterpolationFn = (props: Record<string, any>) => Record<string, any>;

export type SystemPropConfig = {
  property: string;
  styleProperty?: string;
  scale?: keyof Pick<Theme, 'colors' | 'space' | 'opacity'>;
  disabled?: boolean;
  transform?: (value: any) => Record<string, any> | undefined;
};

export type SystemProp = {
  (props: Record<string, any>): Record<string, any>;
  propName: string;
  disabled: boolean;
};
