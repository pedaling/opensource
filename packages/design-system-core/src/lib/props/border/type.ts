import type { ResponsiveValue } from '../../types';

export type BorderProps = {
  borderWidth?: ResponsiveValue<number>;
  borderStyle?: ResponsiveValue<'solid' | 'none'>;
  borderColor?: ResponsiveValue<string>;
  borderTopWidth?: ResponsiveValue<number>;
  borderTopStyle?: ResponsiveValue<'solid' | 'none'>;
  borderTopColor?: ResponsiveValue<string>;
  borderRightWidth?: ResponsiveValue<number>;
  borderRightStyle?: ResponsiveValue<'solid' | 'none'>;
  borderRightColor?: ResponsiveValue<string>;
  borderBottomWidth?: ResponsiveValue<number>;
  borderBottomStyle?: ResponsiveValue<'solid' | 'none'>;
  borderBottomColor?: ResponsiveValue<string>;
  borderLeftWidth?: ResponsiveValue<number>;
  borderLeftStyle?: ResponsiveValue<'solid' | 'none'>;
  borderLeftColor?: ResponsiveValue<string>;
  borderRadius?: ResponsiveValue<number | string>;
  borderTopLeftRadius?: ResponsiveValue<number | string>;
  borderTopRightRadius?: ResponsiveValue<number | string>;
  borderBottomLeftRadius?: ResponsiveValue<number | string>;
  borderBottomRightRadius?: ResponsiveValue<number | string>;
  outlineWidth?: ResponsiveValue<number>;
  outlineStyle?: ResponsiveValue<'solid' | 'none'>;
  outlineColor?: ResponsiveValue<string>;
  outlineOffset?: ResponsiveValue<number>;
};
