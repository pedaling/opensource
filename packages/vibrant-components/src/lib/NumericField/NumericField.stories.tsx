import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { NumericField } from './NumericField';

export default {
  title: 'NumericField',
  component: NumericField,
} as ComponentMeta<typeof NumericField>;

export const Basic: ComponentStory<typeof NumericField> = props => <NumericField {...props} />;
