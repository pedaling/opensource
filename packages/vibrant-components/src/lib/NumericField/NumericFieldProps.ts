import { propVariant, withVariation } from '@vibrant-ui/core';
import type { BaseInputProps } from '../../types';

export type NumericFieldProps = BaseInputProps<number> & {
  placeholder?: number;
  min?: number;
  max?: number;
};

export const withNumericFieldVariation = withVariation<NumericFieldProps>('NumericField')(
  propVariant({
    props: [
      {
        name: 'disabled',
        keep: true,
        default: false,
      },
    ],
    variants: {
      false: {
        color: 'onView1',
        placeholderColor: 'onView2',
        backgroundColor: 'surface3',
      },
      true: {
        color: 'onView3',
        placeholderColor: 'onView3',
        backgroundColor: 'disable',
      },
    },
  })
);
