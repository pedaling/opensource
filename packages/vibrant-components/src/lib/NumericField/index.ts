export { NumericField } from './NumericField';
export type { NumericFieldProps } from './NumericFieldProps';
