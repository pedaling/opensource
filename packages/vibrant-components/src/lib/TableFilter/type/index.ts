export type StringFilterOperator = 'contains' | 'empty' | 'equals' | 'notContains' | 'notEmpty' | 'notEquals';

export type StringFilter = {
  value: string;
  operator: StringFilterOperator[];
};

export type Filter = StringFilter & { dataKey: string };
