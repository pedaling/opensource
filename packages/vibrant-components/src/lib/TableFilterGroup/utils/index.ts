import type { DateFilterOperator, MultiSelectFilterOperator, Option, StringFilterOperator } from '../types';

export const isValueRequiredOperator = (op: string) => op === 'empty' || op === 'notEmpty';

export const isStringFilterValid = (filter: { value: string; operator: StringFilterOperator }) =>
  Boolean(filter.value) || isValueRequiredOperator(filter.operator);

export const isDateFilterValid = (filter: { value: Date[]; operator: DateFilterOperator }) => {
  if (isValueRequiredOperator(filter.operator)) return true;

  if (filter.operator === 'between') return filter.value.length >= 2;

  return filter.value.length >= 1;
};

export const isMultiSelectFilterValid = (filter: { value: Option['value'][]; operator: MultiSelectFilterOperator }) =>
  Boolean(filter.value.length > 0) || isValueRequiredOperator(filter.operator);
