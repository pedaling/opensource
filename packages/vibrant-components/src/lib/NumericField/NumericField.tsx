import { useCallback, useEffect, useRef, useState } from 'react';
import { Box } from '@vibrant-ui/core';
import { Input } from '../Input';
import { OperatorButton } from '../OperatorButton';
import { withNumericFieldVariation } from './NumericFieldProps';

export const NumericField = withNumericFieldVariation(
  ({
    innerRef,
    disabled,
    color,
    defaultValue,
    id,
    max,
    placeholder,
    onValueChange,
    backgroundColor,
    min,
    tabIndex,
    ...restProps
  }) => {
    const inputRef = useRef<HTMLInputElement>();
    const [inputValue, setInputValue] = useState<number | undefined>(defaultValue);

    useEffect(() => {
      setInputValue(defaultValue);
    }, [defaultValue]);

    const updateInputValue = useCallback(
      (nextInputValue: number) => {
        if (min !== undefined && min > nextInputValue) {
          setInputValue(min);

          return;
        }

        if (max !== undefined && max < nextInputValue) {
          setInputValue(max);

          return;
        }

        setInputValue(nextInputValue);
      },
      [max, min]
    );

    useEffect(() => {
      if (
        inputValue === undefined ||
        (min !== undefined && min > inputValue) ||
        (max !== undefined && max < inputValue)
      ) {
        return;
      }

      onValueChange?.(inputValue);
    }, [inputValue, max, min, onValueChange]);

    return (
      <Box position="relative" width={128} height={38} {...restProps}>
        <Input
          ref={(ref: HTMLInputElement) => {
            inputRef.current = ref;
            if (!innerRef) {
              return;
            }

            if (typeof innerRef === 'function') {
              innerRef(ref);

              return;
            }

            innerRef.current = ref;
          }}
          type="number"
          px={38}
          value={inputValue?.toString() ?? ''}
          width="100%"
          backgroundColor={backgroundColor}
          height="100%"
          textAlign="center"
          borderWidth={1}
          borderColor="outline1"
          borderStyle="solid"
          borderRadius={2}
          placeholder={placeholder?.toString()}
          disabled={disabled}
          hideInputSpinButton={true}
          pseudoFocus={{
            borderColor: 'outlineNeutral',
            borderWidth: 1,
            outlineWidth: 0,
          }}
          color={color}
          id={id}
          max={max}
          onValueChange={value => setInputValue(value ? parseInt(value, 10) : undefined)}
          onBlur={() => updateInputValue(inputValue ?? min ?? 0)}
          min={min}
          tabIndex={tabIndex}
        />
        <Box position="absolute" top={4} left={4} bottom={4}>
          <OperatorButton
            operator="minus"
            disabled={(min !== undefined && min >= (inputValue ?? 0)) || disabled}
            onClick={() =>
              setInputValue((value = placeholder ?? min ?? 0) =>
                min !== undefined ? Math.max(value - 1, min) : value - 1
              )
            }
          />
        </Box>
        <Box position="absolute" top={4} right={4} bottom={4}>
          <OperatorButton
            operator="plus"
            disabled={(max !== undefined && max <= (inputValue ?? 0)) || disabled}
            onClick={() =>
              setInputValue((value = placeholder ?? min ?? 0) =>
                max !== undefined ? Math.min(value + 1, max) : value + 1
              )
            }
          />
        </Box>
      </Box>
    );
  }
);
