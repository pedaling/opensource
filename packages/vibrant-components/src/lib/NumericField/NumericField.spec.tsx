import { fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import type { ReactRenderer } from '@vibrant-ui/utils/testing-web';
import { createReactRenderer } from '@vibrant-ui/utils/testing-web';
import { NumericField } from './NumericField';

describe('<NumericField />', () => {
  const { render } = createReactRenderer();

  let renderer: ReactRenderer;
  let element: HTMLElement;
  let mockOnValueChange: jest.Mock;

  beforeEach(() => {
    mockOnValueChange = jest.fn();
  });

  describe('when NumericField rendered', () => {
    let plusButtonElement: HTMLButtonElement;
    let minusButtonElement: HTMLButtonElement;

    beforeEach(() => {
      renderer = render(<NumericField data-testid="numericField" onValueChange={mockOnValueChange} />);

      element = renderer.getByTestId('numericField');

      [minusButtonElement, plusButtonElement] = Array.from(element.querySelectorAll('button')) as HTMLButtonElement[];
    });

    it('input value is empty', () => {
      expect(element.querySelector('input')?.value).toBe('');
    });

    it('onValueChange not called', () => {
      expect(mockOnValueChange).not.toBeCalled();
    });

    describe('when plus button clicked', () => {
      beforeEach(async () => {
        await waitFor(() => userEvent.click(plusButtonElement));
      });

      it('onChangeValue called with 1', () => {
        expect(mockOnValueChange).toBeCalledWith(expect.objectContaining({ value: 1 }));
      });
    });

    describe('when minus button clicked', () => {
      beforeEach(async () => {
        await waitFor(() => userEvent.click(minusButtonElement));
      });

      it('onChangeValue called with -1', () => {
        expect(mockOnValueChange).toBeCalledWith(expect.objectContaining({ value: -1 }));
      });
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when NumberField with min and max rendered', () => {
    let plusButtonElement: HTMLButtonElement;
    let minusButtonElement: HTMLButtonElement;

    beforeEach(() => {
      renderer = render(<NumericField data-testid="numericField" min={3} max={5} onValueChange={mockOnValueChange} />);

      element = renderer.getByTestId('numericField');

      [minusButtonElement, plusButtonElement] = Array.from(element.querySelectorAll('button')) as HTMLButtonElement[];
    });

    it('minus button is disabled', () => {
      expect(minusButtonElement.disabled).toBe(true);
    });

    it('plus button is enabled', () => {
      expect(plusButtonElement.disabled).toBe(false);
    });

    describe('when plus button clicked', () => {
      beforeEach(async () => {
        await waitFor(() => userEvent.click(plusButtonElement));
      });

      it('onValueChange called with 4', () => {
        expect(mockOnValueChange).toBeCalledWith(expect.objectContaining({ value: 4 }));
      });

      it('minus button is enabled', () => {
        expect(minusButtonElement.disabled).toBe(false);
      });
    });

    describe('when press 1 to input', () => {
      beforeEach(done => {
        const input = element.querySelector('input');

        if (!input) {
          done.fail();

          return;
        }

        waitFor(() => userEvent.type(input, '1')).then(done);
      });

      it('onValueChange not called', () => {
        expect(mockOnValueChange).not.toBeCalled();
      });

      describe('when blur', () => {
        beforeEach(async () => {
          await waitFor(() => fireEvent.blur(element.querySelector('input')!));
        });

        it('onValueChange called with min value', () => {
          expect(mockOnValueChange).toBeCalledWith(expect.objectContaining({ value: 3 }));
        });
      });
    });
  });
});
