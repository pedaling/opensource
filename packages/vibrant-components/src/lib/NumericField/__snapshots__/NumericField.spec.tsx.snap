// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumericField /> when NumericField rendered match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  width: 128px;
  height: 38px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  background-color: #ffffff;
  border-width: 1px;
  color: #0c0c0c;
  -webkit-text-fill-color: #0c0c0c;
  text-fill-color: #0c0c0c;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.125rem;
  width: 100%;
  padding: 0;
  padding-left: 38px;
  padding-right: 38px;
  height: 100%;
  text-align: center;
  border-color: #00000026;
  border-style: solid;
  border-radius: 2px;
  -moz-appearance: textfield;
}

.emotion-1::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.emotion-1::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #5b5b5b;
  -webkit-text-fill-color: #5b5b5b;
  text-fill-color: #5b5b5b;
}

.emotion-1::-moz-placeholder {
  color: #5b5b5b;
  -webkit-text-fill-color: #5b5b5b;
  text-fill-color: #5b5b5b;
}

.emotion-1:-ms-input-placeholder {
  color: #5b5b5b;
  -webkit-text-fill-color: #5b5b5b;
  text-fill-color: #5b5b5b;
}

.emotion-1::placeholder {
  color: #5b5b5b;
  -webkit-text-fill-color: #5b5b5b;
  text-fill-color: #5b5b5b;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: absolute;
  top: 4px;
  left: 4px;
  bottom: 4px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  cursor: pointer;
  background-color: #0000000d;
  border-width: 0;
  padding: 0;
  overflow: hidden;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  z-index: 0;
  border-radius: 2px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: absolute;
  z-index: -1;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  background-color: #0c0c0c;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  width: auto;
  height: auto;
  padding: 7px;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  width: 16px;
  height: 16px;
  fill: #0c0c0c;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: absolute;
  top: 4px;
  right: 4px;
  bottom: 4px;
}

<div>
  <div
    class="emotion-0"
    data-testid="numericField"
  >
    <input
      autocomplete="off"
      class="emotion-1"
      inputmode="numeric"
      type="text"
      value=""
    />
    <div
      class="emotion-2"
    >
      <button
        class="emotion-3"
        type="button"
      >
        <span
          class="emotion-4"
          style="opacity: 0;"
        />
        <span
          class="emotion-5"
          style="opacity: 1;"
        >
          <svg
            class="emotion-6"
            viewBox="0 0 24 24"
          >
            <path
              class="emotion-7"
              d="M21.25 13.25H2.75c-.15 0-.25-.1-.25-.25v-2c0-.15.1-.25.25-.25h18.5c.15 0 .25.1.25.25v2c0 .15-.1.25-.25.25Z"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="emotion-8"
    >
      <button
        class="emotion-3"
        type="button"
      >
        <span
          class="emotion-4"
          style="opacity: 0;"
        />
        <span
          class="emotion-5"
          style="opacity: 1;"
        >
          <svg
            class="emotion-6"
            viewBox="0 0 24 24"
          >
            <path
              class="emotion-7"
              d="M21.25 10.75h-8v-8c0-.15-.1-.25-.25-.25h-2c-.15 0-.25.1-.25.25v8h-8c-.15 0-.25.1-.25.25v2c0 .15.1.25.25.25h8v8c0 .15.1.25.25.25h2c.15 0 .25-.1.25-.25v-8h8c.15 0 .25-.1.25-.25v-2c0-.15-.1-.25-.25-.25Z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
